libs/dyn/pattern.go

```diff
@@ -32,6 +32,21 @@ func NewPatternFromPath(p Path) Pattern {
 	return cs
 }
 
+// Split <pattern>.<string_key> into <pattern> and <string_key>
+// The last component must be dyn.Key() and there must be at least two components.
+func (p Pattern) Split() (Pattern, string) {
+	if len(p) <= 1 {
+		return nil, ""
+	}
+	parent := p[:len(p)-1]
+	leaf := p[len(p)-1]
+	pc, ok := leaf.(pathComponent)
+	if !ok {
+		return nil, ""
+	}
+	return parent, pc.Key()
```

Comment:
It is also invalid if the component is an index, per the description of the function.

Or alternatively, it should return the path component itself, and the caller determines if it is a key or index component. This matches the spirit of `Split` better, IMO, though it puts the burden on the caller to make the assertion that it is a key.